File: src/lorecraft/protocol/envelope.py
"""Command ingress envelope and outcome — Python mirror of `envelope.rs`.

ID fields are plain `str` aliases (matching the Rust `#[serde(transparent)]`
newtypes' bare-string JSON), so no wrapping is needed on this side.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum

from lorecraft.protocol.effects import Effect
from lorecraft.protocol.messages import OutboundMessage

# ID aliases — bare strings on the wire, matching Rust's transparent newtypes.
WorldId = str
ActorId = str
PlayerId = str
SessionId = str
CommandId = str


class OutcomeStatus(str, Enum):
    """Terminal status of a command's execution.

    Subclasses ``str`` so a value serializes as its plain variant name (matching
    the Rust externally-tagged unit-enum wire shape, e.g. ``"Executed"``).
    """

    EXECUTED = "Executed"
    BLOCKED = "Blocked"
    FAILED = "Failed"
    TIMED_OUT = "TimedOut"


@dataclass(frozen=True, slots=True)
class Diagnostic:
    """A diagnostic annotation attached to an outcome or script result."""

    level: str
    message: str


@dataclass(frozen=True, slots=True)
class CommandEnvelope:
    """A command as admitted to the engine, with routing/idempotency metadata."""

    protocol_version: int
    world_id: WorldId
    actor_id: ActorId
    player_id: PlayerId
    session_id: SessionId
    command_id: CommandId
    receive_sequence: int
    deadline_ms: int
    raw: str


@dataclass(frozen=True, slots=True)
class CommandOutcome:
    """The authoritative result of executing a ``CommandEnvelope``."""

    command_id: CommandId
    status: OutcomeStatus
    commit_sequence: int | None = None
    messages: list[OutboundMessage] = field(default_factory=list)
    applied_effects: list[Effect] = field(default_factory=list)
    diagnostics: list[Diagnostic] = field(default_factory=list)

File: rust/crates/lorecraft-protocol/src/envelope.rs
//! Command ingress envelope and execution outcome.

use serde::{Deserialize, Serialize};

use crate::effects::Effect;
use crate::ids::{ActorId, CommandId, PlayerId, SessionId, WorldId};
use crate::messages::OutboundMessage;

/// A command as admitted to the engine, carrying all routing/idempotency metadata.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct CommandEnvelope {
    /// Protocol version the envelope was produced under.
    pub protocol_version: u32,
    /// World the command targets.
    pub world_id: WorldId,
    /// Actor issuing the command.
    pub actor_id: ActorId,
    /// Player account behind the actor.
    pub player_id: PlayerId,
    /// Session/connection the command arrived on.
    pub session_id: SessionId,
    /// Idempotency key for this command.
    pub command_id: CommandId,
    /// Monotonic admission/client sequence number.
    pub receive_sequence: u64,
    /// Monotonic execution budget in milliseconds.
    pub deadline_ms: u64,
    /// Raw command line as typed by the player.
    pub raw: String,
}

/// Terminal status of a command's execution.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub enum OutcomeStatus {
    /// Command ran and committed successfully.
    Executed,
    /// Command was rejected before mutation (a rule/precondition blocked it).
    Blocked,
    /// Command failed during execution.
    Failed,
    /// Command exceeded its execution deadline.
    TimedOut,
}

/// A diagnostic annotation attached to an outcome or script result.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Diagnostic {
    /// Severity/category label, e.g. `"info"`, `"warning"`, `"error"`.
    pub level: String,
    /// Human-readable diagnostic message.
    pub message: String,
}

/// The authoritative result of executing a [`CommandEnvelope`].
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct CommandOutcome {
    /// Command this outcome corresponds to.
    pub command_id: CommandId,
    /// Terminal status.
    pub status: OutcomeStatus,
    /// Commit sequence assigned if the command committed state, else `None`.
    pub commit_sequence: Option<u64>,
    /// Messages to deliver to the client, in order.
    pub messages: Vec<OutboundMessage>,
    /// Effects the engine actually applied.
    pub applied_effects: Vec<Effect>,
    /// Diagnostics produced during execution.
    pub diagnostics: Vec<Diagnostic>,
}
